### PR TITLE
Use System preference for light dark theme

### LINKED ISSUE #8729: Option to use System Preference for Light/Dark Theme (Win/Lin/OSX)
**Is your suggestion for improvement related to a problem? Please describe.**

I would appreciate a way to automatically apply either a light or dark theme whenever I switch my light/dark theme in my MacOS system preferences. Sometimes I am using an eink monitor and I prefer light themes because they are easier to read on eink IMO, and other times I am on an LCD panel monitor and do not want to hurt my eyes with a light mode (I want dark mode here).

**Describe the solution you'd like**

Could a radio button be added to the Appearance section of the settings dialog that says "Use System Default"?

**Additional context**

I would prefer a solution that does not require a program restart, but this may be easier or just a good first step to aim for.

Thank you to everyone who has contributed to this software!

### PR DESCRIPTION
**Pull request by EthanYifanJu**
Resolve #8729 
Based on the work of #9263 

In the Preference -> General -> Appearance -> Visual Theme drop-down list, I added one more option called "Use System Preference", which will change between light/dark themes according to current OS system settings. [jSystemThemeDetector](https://github.com/Dansoftowner/jSystemThemeDetector) is used to achieve this goal.

Screenshot 2023-10-28 at 12 29 26

Manually test the outcome on the MacOS. As expected, the theme will change according to the Mac appearance setting.

https://github.com/JabRef/jabref/assets/111887946/941ed624-960a-428b-8b04-f8b6b86e7a25

Manually test the outcome on the Windows. As expected, the theme will change according to the Windows Colour setting (Sorry I can't change the language into English for this Windows device. The three options are Light/Dark/Custom in the drop-down list). This device is relatively old, so I assume the delay is normal when changing themes.

https://github.com/JabRef/jabref/assets/111887946/41b5801b-3160-4001-8496-4f977c2b096d

### Mandatory checks

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -257,6 +257,8 @@ dependencies {
     rewrite("org.openrewrite.recipe:rewrite-logging-frameworks")
     rewrite("org.openrewrite.recipe:rewrite-testing-frameworks")
     rewrite("org.openrewrite.recipe:rewrite-migrate-java")
+
+    implementation 'com.github.Dansoftowner:jSystemThemeDetector:3.6'
```
Comment (koppor) [inline build.gradle:261]: Please try to keep the order in `build.gradle` --> move it up to the other implementation calls.

File: licenses/com.jthemedetector.OsThemeDetector [hunk 2]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (koppor) [inline licenses/com.jthemedetector.OsThemeDetector:1]: No need to add that - we do not keep license files

File: src/main/java/org/jabref/gui/JabRefGUI.java [hunk 3]
```diff
@@ -145,6 +147,31 @@ private void openWindow(Stage mainStage) {
         root.getChildren().add(JabRefGUI.mainFrame);
 
         Scene scene = new Scene(root, 800, 800);
+        // Call the OsThemeDetector from jSystemThemeDetector library (https://github.com/Dansoftowner/jSystemThemeDetector)
```
Comment (koppor) [inline src/main/java/org/jabref/gui/JabRefGUI.java:150]: Remove this comment - it is stating the same as in the next line

File: src/main/java/org/jabref/gui/JabRefGUI.java [hunk 4]
```diff
@@ -145,6 +147,31 @@ private void openWindow(Stage mainStage) {
...
+        final OsThemeDetector detector = OsThemeDetector.getDetector();
+        // If OS theme is light or dark, JabRef UI Theme will correspond to it. If JabRef is then closed and user modifies their
+        // OS theme settings, the code below will make the necessary theme switch IF automatic detection is enabled by user in
+        // JabRef Appearance Preferences.
+        if (preferencesService.getWorkspacePreferences().automaticThemeDetectionFlag().getValue()) {
+            final boolean isDarkThemeUsed = detector.isDark();
+            if (isDarkThemeUsed) {
+                // The OS uses a dark theme
```
Comment (koppor) [inline src/main/java/org/jabref/gui/JabRefGUI.java:158]: Remove this comment - the `if` the line above says the same

File: src/main/java/org/jabref/gui/JabRefGUI.java [hunk 5]
```diff
@@ -145,6 +147,31 @@ private void openWindow(Stage mainStage) {
...
+                preferencesService.getWorkspacePreferences().setTheme(Theme.dark());
+            } else {
+                // The OS uses a light theme
```
Comment (koppor) [inline src/main/java/org/jabref/gui/JabRefGUI.java:161]: Remove this comment - the `if` the line above says the same

File: src/main/java/org/jabref/gui/JabRefGUI.java [hunk 6]
```diff
@@ -145,6 +147,31 @@ private void openWindow(Stage mainStage) {
         root.getChildren().add(JabRefGUI.mainFrame);
 
         Scene scene = new Scene(root, 800, 800);
+        // Call the OsThemeDetector from jSystemThemeDetector library (https://github.com/Dansoftowner/jSystemThemeDetector)
+        final OsThemeDetector detector = OsThemeDetector.getDetector();
+        // If OS theme is light or dark, JabRef UI Theme will correspond to it. If JabRef is then closed and user modifies their
+        // OS theme settings, the code below will make the necessary theme switch IF automatic detection is enabled by user in
+        // JabRef Appearance Preferences.
```
Comment (koppor) [inline src/main/java/org/jabref/gui/JabRefGUI.java:154]: Remove this comment - the code is self-explanatory

File: src/main/java/org/jabref/gui/JabRefGUI.java [hunk 7]
```diff
@@ -145,6 +147,31 @@ private void openWindow(Stage mainStage) {
...
+            final boolean isDarkThemeUsed = detector.isDark();
+            if (isDarkThemeUsed) {
+                // The OS uses a dark theme
+                preferencesService.getWorkspacePreferences().setTheme(Theme.dark());
+            } else {
+                // The OS uses a light theme
+                preferencesService.getWorkspacePreferences().setTheme(Theme.light());
+            }
+        }
+        // Listening to changes
```
Comment (koppor) [inline src/main/java/org/jabref/gui/JabRefGUI.java:165]: Remove this comment

File: src/main/java/org/jabref/gui/JabRefGUI.java [hunk 8]
```diff
@@ -145,6 +147,31 @@ private void openWindow(Stage mainStage) {
...
+        detector.registerListener(isDark -> {
+            if (isDark) {
+                // The OS switched to a dark theme
```
Comment (koppor) [inline src/main/java/org/jabref/gui/JabRefGUI.java:168]: Remove this comment

File: src/main/java/org/jabref/gui/JabRefGUI.java [hunk 9]
```diff
@@ -145,6 +147,31 @@ private void openWindow(Stage mainStage) {
...
+                preferencesService.getWorkspacePreferences().setTheme(Theme.dark());
+            } else {
+                // The OS switched to a light theme
```
Comment (koppor) [inline src/main/java/org/jabref/gui/JabRefGUI.java:171]: Remove this comment

File: src/main/java/org/jabref/gui/preferences/general/GeneralTabViewModel.java [hunk 10]
```diff
@@ -125,8 +129,18 @@ public void setValues() {
         // The light theme is in fact the absence of any theme modifying 'base.css'. Another embedded theme like
         // 'dark.css', stored in the classpath, can be introduced in {@link org.jabref.gui.theme.Theme}.
         switch (workspacePreferences.getTheme().getType()) {
-            case DEFAULT -> selectedThemeProperty.setValue(ThemeTypes.LIGHT);
-            case EMBEDDED -> selectedThemeProperty.setValue(ThemeTypes.DARK);
+            case DEFAULT -> {
+                selectedThemeProperty.setValue(ThemeTypes.LIGHT);
```
Comment (koppor) [inline src/main/java/org/jabref/gui/preferences/general/GeneralTabViewModel.java:133]: I don't get, why first LIGHT is set and then SYNC. Why not an `else` branch at the if?

File: src/main/java/org/jabref/gui/preferences/general/GeneralTabViewModel.java [hunk 11]
```diff
@@ -125,8 +129,18 @@ public void setValues() {
...
+                if (workspacePreferences.automaticThemeDetectionFlag().getValue()) {
+                    selectedThemeProperty.setValue(ThemeTypes.SYNC);
+                }
+            }
+            case EMBEDDED -> {
+                selectedThemeProperty.setValue(ThemeTypes.DARK);
```
Comment (koppor) [inline src/main/java/org/jabref/gui/preferences/general/GeneralTabViewModel.java:139]: I don't get, why first DARK is set and then SYNC. Why not an `else` branch at the if?

File: src/main/java/org/jabref/gui/preferences/general/GeneralTabViewModel.java [hunk 12]
```diff
@@ -54,6 +55,9 @@ public class GeneralTabViewModel implements PreferenceTabViewModel {
     private final ReadOnlyListProperty<ThemeTypes> themesListProperty =
             new ReadOnlyListWrapper<>(FXCollections.observableArrayList(ThemeTypes.values()));
     private final ObjectProperty<ThemeTypes> selectedThemeProperty = new SimpleObjectProperty<>();
+
+    private final BooleanProperty flagWhenAutomaticButtonSelected = new SimpleBooleanProperty();
```
Comment (koppor) [inline src/main/java/org/jabref/gui/preferences/general/GeneralTabViewModel.java:59]: Why such a long name? The other names are shorter?

Proposal: 

    automaticSyncActivatedProperty

File: src/main/java/org/jabref/preferences/JabRefPreferences.java [hunk 13]
```diff
@@ -417,6 +417,9 @@ public class JabRefPreferences implements PreferencesService {
     // Helper string
     private static final String USER_HOME = System.getProperty("user.home");
 
+    // System default theme / Auto detect theme constant
```
Comment (koppor) [inline src/main/java/org/jabref/preferences/JabRefPreferences.java:420]: Remove comment

File: src/main/java/org/jabref/preferences/JabRefPreferences.java [hunk 14]
```diff
@@ -787,6 +790,8 @@ private JabRefPreferences() {
 
         // set default theme
         defaults.put(FX_THEME, Theme.BASE_CSS);
+        // Set automatic theme detection OFF.
```
Comment (koppor) [inline src/main/java/org/jabref/preferences/JabRefPreferences.java:793]: Remove that comment

**Review by koppor — DISMISSED** [review]
Some code comments

When I keep "Light", but set in Windows to "Dark", JabRef also gets dark

image

**Comment by calixtus** [comment]
I put some work into this.
grafik
grafik

**Review by Siedlerchr — APPROVED** [review]

**Review by Siedlerchr — APPROVED** [review]

**Pull request merged**